src/System.Buffers.Experimental/System/Buffers/ReadOnlyBytes.cs

```diff
@@ -12,37 +12,42 @@ namespace System.Buffers
     /// </summary>
     public struct ReadOnlyBytes : IReadOnlyBufferList<byte>
     {
-        ReadOnlyMemory<byte> _first;
-        int _totalLength;
+        ReadOnlyMemory<byte> _first; // pointer + index:int + length:int
         IReadOnlyBufferList<byte> _rest;
-        
+        long _totalLength;
+
         static readonly ReadOnlyBytes s_empty = new ReadOnlyBytes(ReadOnlyMemory<byte>.Empty);
 
-        public ReadOnlyBytes(ReadOnlyMemory<byte> first, IReadOnlyBufferList<byte> rest, int length)
+        public ReadOnlyBytes(ReadOnlyMemory<byte> first, IReadOnlyBufferList<byte> rest, long length)
         {
-            Debug.Assert(rest != null || length <= first.Length);
+            if(length < 0) throw new ArgumentOutOfRangeException(nameof(length));
+            if (first.Length + (rest == null ? 0 : rest.Length) < length)
+                throw new ArgumentOutOfRangeException(nameof(length), length, string.Format("{0}+{1}", first.Length, rest.Length));
             _rest = rest;
             _first = first;
             _totalLength = length;
         }
 
         public ReadOnlyBytes(ReadOnlyMemory<byte> first, IReadOnlyBufferList<byte> rest) :
-            this(first, rest, rest==null?first.Length:Unspecified)
-        {
-        }
+            this(first, rest, first.Length + rest.Length)
+        { }
 
         public ReadOnlyBytes(ReadOnlyMemory<byte> buffer) :
             this(buffer, null, buffer.Length)
         { }
 
-        public ReadOnlyBytes(IReadOnlyBufferList<byte> segments, int length) :
-            this(segments.First, segments.Rest, length)
-        { }
-
         public ReadOnlyBytes(IReadOnlyBufferList<byte> segments) :
-            this(segments.First, segments.Rest, Unspecified)
+            this(segments.First, segments.Rest, segments.Length)
         { }
 
+        public ReadOnlyBytes(IReadOnlyBufferList<byte> segments, long length) :
+            this(segments.First, segments.Rest, length)
+        {
+            // TODO: should they be runtime/release checks?
+            Debug.Assert(segments != null);
```

Comment:
Does the base class constructor get called before these asserts? If so, what happens if segments is null and I call `segments.First`?

Reply:
Yeah, I will remove the todo and do all proper checks